packages/cli/src/commands/setup/deploy/providers/aws-serverless.js

```diff
@@ -16,6 +16,7 @@ service: app
 
 plugins:
   - serverless-dotenv-plugin
+  - serverless-api-cloudfront
```

Comment:
🚀

Reply:
Unfortunately, use of this plugin might not be what we want. We currently use [`httpApi` events](https://www.serverless.com/blog/aws-http-api-support) and not `http`; which this [plugin expects.](https://github.com/Droplr/serverless-api-cloudfront#error-troubleshooting) The plugin also does not seem to be updated much, so more of a roll our own approach seems warranted. 

---

Shout out to @ajcwebdev for helping smoke testing this.

Reply:
Ah, bummer. I'm trying to get some help from folks at Serverless. Let's keep this top of the list for them to look at.

Reply:
Hi! Serverless Framework team member here 👋 

Could you clarify why CloudFront is used? Is it used:

- as a CDN for static assets/static website hosting (i.e. in front of S3)
- or as a CDN/cache for HTTP responses returned by Lambda (i.e. in front of API Gateway)

From what I understand it seems to be the second case, but I just want to make sure.

AFAIK here are your option to set up CloudFront with API Gateway in serverless.yml:

- deploy and configure CloudFront via CloudFormation code in `serverless.yml` (in the `resources` section) -> you should expect ~100 lines of YAML (I've done this a lot, this can be painful) -> I can help you with that if you choose to go that way
- or use a plugin, but AFAIK there are not many options - I've worked on a plugin that may help here: https://github.com/getlift/lift

Lift offers several "constructs" to deploy high-level components. One of them is "server-side website": that construct deploys a CloudFront distribution that serves responses from API Gateway.

You can read more about it here: https://github.com/getlift/lift/blob/master/docs/server-side-website.md

One piece of caution: that construct was built with backend frameworks in mind, like Laravel, Ruby on Rails, Django, Symfony… Not really for server-side rendering of frontend websites (even though one user [mentioned that it seemed to work fine too for these use-cases](https://github.com/getlift/lift/discussions/43#discussioncomment-896386)). From what I've read about Redwood, it's the GraphQL API running on Lambda? If that's the case, then it could work well here.

Also that leads me to one question: why use CloudFront in front of the GraphQL API?

By the way we also have one for static websites served with CloudFront + S3: https://github.com/getlift/lift/blob/master/docs/static-website.md (but it may not be what you need here).

Reply:
Hi @mnapoli Thanks for jumping in here. Redwood has two considerations during deployment:
1. API functions
2. Web assets

We use a Jamstack-like deployment flow, which you can read about in the first two sections of this Doc:
- https://redwoodjs.com/docs/deploy

Our goal here is to deploy:
1. API Functions --> AWS Lambdas
2. Web assets --> CDN

The latter should be Cloudfront, correct?

## Bonus Time: Caching
Lastly, Redwood's GraphQL API Function is "cache-ready" out of the box: either mem or DB (eg. Redis). For serverless deployments, we normally reach for Redis, which might not be the only option on AWS Lambdas.

It's not a hard requirement for this PR, but it would 100% be something that gets people excited to use this Serverless setup. Any immediate reactions/ideas?

Reply:
Thanks, that's clearer. I see 2 scenarios:

Scenario A (simplest):

- website.com -> CloudFront -> S3
- api.website.com (i.e. another domain) -> API Gateway -> Lambda

In that case, [Lift's static website](https://github.com/getlift/lift/blob/master/docs/static-website.md) might be perfect here. If you don't want to use the Lift plugin, it is also perfectly possible to write the CloudFormation code manually (or [use "eject" to generate that code](https://github.com/getlift/lift#ejecting)). Note that Lift provides a few helpers as well to quickly deploy asset changes, and clear the CDN cache on deployment.

You get CDN for the static website, but CloudFront would not cover the API. One could use API Gateway's caching feature though (API Gateway has a caching system).

Scenario B:

- website.com/* -> CloudFront
    - /* -> S3
    - /api/* -> API Gateway -> Lambda

In other words, all requests go through CloudFront. Then CloudFront routes based on the path. CloudFront caching could then apply to web assets and API responses. Watch out: this scenario also involves extra CloudFront costs, because all API calls go through CloudFront.

In that case, you may need to write the CloudFront configuration manually.

Reply:
@mnapoli Lift's `static-website` construct is great, thanks for pointing me in this direction!

If we did decide to go with scenario B, do you know of a similar construct for spinning up that setup, or where I might start looking for CDK snippets/constructs? I'd like to avoid manual CloudFront if possible.

Reply:
@virtuoushub For scenario B I'm wondering if https://github.com/getlift/lift/blob/master/docs/server-side-website.md could actually be a solution here 🤔 

Maybe a config like this would be worth testing:

```yaml
service: my-app
provider:
    name: aws

functions:
    function1:
        handler: home.handler
        events:
            -   httpApi: 'GET /.redwood/functions/function1/'
    # ...

constructs:
    website:
        type: server-side-website
        assets:
            '/index.html': dir-of-static-website/index.html
            '/css/*': dir-of-static-website/css
            '/js/*': dir-of-static-website/js

plugins:
    - serverless-lift
```

The downsides:

- the name of that construct (`server-side-website`) isn't really the best name here, but it's just a name 🤷 
- with that construct, since `/*` is directed by default to API URLs, you need to specify explicitly every root file in the `constructs.website.assets` section (you cannot do `'/*': dir-of-static-website/`, or at least I don't expect it to work)